### PR TITLE
Feature/dg 979 support classification

### PR DESCRIPTION
**Pull request by BloodAxe**
* Adds support for classification task on public API level
* Adds simple features (class distribution)
* Improve `ask_user` method to use ipython widgets instead of stdin when launched from notebook to allow users to provide input

### CODE REVIEW

**Comment by shaydeci** [comment]
> * Adds support for classification task on public API level
> * Adds simple features (class distribution)
> * Improve `ask_user` method to use ipython widgets instead of stdin when launched from notebook to allow users to provide input

In SG we kept colab links rather then having the opynb files inside the repo, I think its better that way (in case changes are introduced we will always need to update in 2 places...)

**Comment by BloodAxe** [comment]
> > * Adds support for classification task on public API level
> > * Adds simple features (class distribution)
> > * Improve `ask_user` method to use ipython widgets instead of stdin when launched from notebook to allow users to provide input
> 
> In SG we kept colab links rather then having the ipynb files inside the repo, I think its better that way (in case changes are introduced we will always need to update in 2 places...)

Well I can argue that this break "atomic change" concept, that you have to make a DF release first, then go over all colab files and update them by hand. From my perspective having them under git is easier to maintain. These are meant to be "quick usage examples" that people can quickly refer to.
Maybe for deep dive colabs as we did for YOLO-NAS it make sense to store them elsewhere...
But honestly - location of notebook files is the smallest question I have about this PR ):

File: src/data_gradients/managers/abstract_manager.py [hunk 1]
```diff
@@ -63,6 +66,26 @@ def __init__(
...
         self.batches_early_stop = batches_early_stop
+
+        # Check if train_data and val_data are DataLoader objects.
+        # If not, try to convert them to DataLoader objects.
+        # Generally this should work fine, with only caveat that detection datasets may require custom collate_fn.
+        # However since we are using bs=1 this should not be a problem, but just in case we try to get a singe batch
+        # from the dataset and see if it works.
+        if not isinstance(train_data, DataLoader):
+            try:
+                next(iter(DataLoader(train_data)))
```
Comment (Louis-Dupont) [inline src/data_gradients/managers/abstract_manager.py:77]: `next(iter(DataLoader(train_data)))` This is to check if we can properly iterate over the DataLoader ?

Comment (Louis-Dupont) [reply src/data_gradients/managers/abstract_manager.py:77]: If `train_data` is a generator, this would drop the first element. Maybe it's an edge case that we can live with, but it's not ideal if the user runs on the whole dataset.
The other edge case we discussed was if the user passes an iterable that returns batches but is not a DataLoader.

Do we still want to handle it like that ?
The alternative is to handle the batch/not batch at the batch_processor level, like in segmentation.

File: src/data_gradients/feature_extractors/classification/class_distribution.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        all_class_names = df_class_count["class_name"].unique()
+
+        # Height of the plot is proportional to the number of classes
+        n_unique = len(all_class_names)
+        figsize_x = 10
+        figsize_y = min(max(6, int(n_unique * 0.3)), 175)
+
+        plot_options = BarPlotOptions(
+            x_label_key="n_appearance",
+            x_label_name="Class support (# of samples)",
```
Comment (Louis-Dupont) [inline src/data_gradients/feature_extractors/classification/class_distribution.py:43]: Not sure if I'm the only one, but I feel like `#` in `Class support (# of samples)` looks kinda dirty in the context of a report

File: src/data_gradients/feature_extractors/classification/class_distribution.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+from data_gradients.common.registry.registry import register_feature_extractor
+from data_gradients.feature_extractors.abstract_feature_extractor import Feature
+from data_gradients.utils.data_classes.data_samples import ClassificationSample
+from data_gradients.visualize.seaborn_renderer import BarPlotOptions
+from data_gradients.feature_extractors.abstract_feature_extractor import AbstractFeatureExtractor
+
+
+@register_feature_extractor()
+class ClassificationClassDistribution(AbstractFeatureExtractor):
```
Comment (Louis-Dupont) [inline src/data_gradients/feature_extractors/classification/class_distribution.py:13]: The name is not consistent with Segmentation and Detection. (`SegmentationClassFrequency` and `DetectionClassFrequency`)

File: src/data_gradients/feature_extractors/classification/class_distribution.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        df = pd.DataFrame(self.data)
+
+        # Include ("class_name", "split", "n_appearance")
+        df_class_count = df.groupby(["class_name", "class_id", "split"]).size().reset_index(name="n_appearance")
+        all_class_names = df_class_count["class_name"].unique()
+
+        # Height of the plot is proportional to the number of classes
+        n_unique = len(all_class_names)
+        figsize_x = 10
+        figsize_y = min(max(6, int(n_unique * 0.3)), 175)
```
Comment (Louis-Dupont) [inline src/data_gradients/feature_extractors/classification/class_distribution.py:39]: In segmentation and classification, we have an option to filter to only keep topk

```python
df_class_count = self.value_extractor.select(df=df_class_count, id_col="class_id", split_col="split", value_col="frequency")
```

I think it would be good to add this option (great for cases with huge number of classes)

File: src/data_gradients/feature_extractors/classification/class_distribution.py (hunk 3 above)
Comment (Louis-Dupont) [reply src/data_gradients/feature_extractors/classification/class_distribution.py:13]: But I also notice that the variables shown are different.
Here you show the count, and in the seg/detection it's the frequency.
We moved from count to frequency because it helps to compare train/val sets distribution gap

**Review by Louis-Dupont — COMMENTED** [review]
Looks good, I left some comments inline but nothing too important.

As a bonus, if you can add a minimalistic end-to-end test with a dummy dataset (which can be built on the fly with random images and only a couple of random classes.
We don't have such tests for segmentation and detection but I think we should asap.

File: src/data_gradients/feature_extractors/classification/class_distribution.py (hunk 3 above)
Comment (BloodAxe) [reply src/data_gradients/feature_extractors/classification/class_distribution.py]: Fixed in 8530cb21cca2851d3fadfdc0f57f82362db43853

File: src/data_gradients/feature_extractors/classification/class_distribution.py (hunk 4 above)
Comment (BloodAxe) [reply src/data_gradients/feature_extractors/classification/class_distribution.py]: Fixed in 8530cb21cca2851d3fadfdc0f57f82362db43853 as well.
However top_k argument is made None by default. 
I beleive what we eventually want to do is have top_k is argument in analysis manager.
So user can tune top_k from the API.

**Review by ofrimasad — COMMENTED** [review]
a few minor comments

File: src/data_gradients/batch_processors/formatters/classification.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        """
+
+        images = ensure_channel_first(images, n_image_channels=self.n_image_channels)
+        images = ensure_images_shape(images, n_image_channels=self.n_image_channels)
+        labels = self.ensure_labels_shape(images=images, labels=labels)
+
+        if 0 <= images.min() and images.max() <= 1:
+            images *= 255
+            images = images.to(torch.uint8)
+
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/data_gradients/batch_processors/formatters/classification.py:62]: Fixed in 43c4f16cddd06dfb801da4414cca7e2efb3bf7d1

File: src/data_gradients/assets/html/basic_info_fe_classification.html
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/data_gradients/assets/html/basic_info_fe_classification.html:1]: I think having a dedicated feature extractor class and a template for a task is better and more explicit. I feel this this is a right way to do it, since otherwise your "one template fits all" will be overloaded with a bunch of `if` statements checking whether certain fields are filled. Also in makes `aggregate` logic more complicated if you have to take into consideration whether some fields are empty because there were no samples or that fields is irrelevant for the task. And last but not least in `update` method this requires you to use `isinstance(sample, ...)` to infer sample type, which also require one to keep in mind this place when adding a new task.

File: src/data_gradients/batch_processors/formatters/classification.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        """
+
+        images = ensure_channel_first(images, n_image_channels=self.n_image_channels)
+        images = ensure_images_shape(images, n_image_channels=self.n_image_channels)
+        labels = self.ensure_labels_shape(images=images, labels=labels)
+
+        if 0 <= images.min() and images.max() <= 1:
+            images *= 255
+            images = images.to(torch.uint8)
+        elif images.min() < 0:  # images were normalized with some unknown mean and std
```
Comment (shaydeci) [inline src/data_gradients/batch_processors/formatters/classification.py:57]: This won't get us the original image though...but maybe its the best we can do ?
I am wondering if in this case, we would rather have a visualizable image that has different pixel values then something else.

Just thinking out loud here, whether users would be confused by the result image from this.

File: requirements.txt [hunk 7]
```diff
@@ -17,3 +17,4 @@ seaborn
 xhtml2pdf
 jinja2
 imagededup
+jupyter-ui-poll==0.2.2
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply requirements.txt]: That is a good point. I had the intention to try and see whether it works and then bring source inside DG on the code level, but forgot 🤦 
Fixed in 6693126efd22a3b1cfafd13aed045b92de1a162a

**Review by shaydeci — CHANGES_REQUESTED** [review]

File: src/data_gradients/managers/abstract_manager.py (hunk 1 above)
Comment (BloodAxe) [reply src/data_gradients/managers/abstract_manager.py:77]: That's fair. I suggest to ignore this edge case for now 🙈

File: src/data_gradients/batch_processors/formatters/classification.py (hunk 6 above)
Comment (BloodAxe) [reply src/data_gradients/batch_processors/formatters/classification.py:57]: Ok, I'm continuing thinking out loud as well. Why do we do denormalization in the first place?
Imagine I have my dataset transforms that do normalization and I want to see colors distribution after normalization (to see how close they are to zero mean / unit std). But instead we are forcing users to 8-bit images. There are uint16 images, there are SAR images with full fp32 range.
They only way we need denormalization, IMO, is visualization of samples. And here, if no mean/std or custom denormalization function is provided we can do scaling to 0..255 range with a warning message that results may look odd.

Comment (shaydeci) [reply src/data_gradients/batch_processors/formatters/classification.py:57]: Warning sounds good

**Comment by Louis-Dupont** [comment]
Looks like the notebook still includes this plot
image

File: src/data_gradients/managers/classification_manager.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    """Implementation of analysys manager for image classification task.
+    Definition of task name, task-related preprocessor and parsing related configuration file
+    """
+
+    def __init__(
+        self,
+        *,
+        report_title: str,
+        train_data: Iterable,
+        val_data: Optional[Iterable] = None,
```
Comment (Louis-Dupont) [inline src/data_gradients/managers/classification_manager.py:22]: Actually, do we really support `val_data=None`? With other tasks we don't fully support it, so if it's also the case for classification let's remove `Optional`

Comment (BloodAxe) [reply src/data_gradients/managers/classification_manager.py:22]: We do. For classification task this is supported.

**Review by Louis-Dupont — APPROVED** [review]
LGTM

**Comment by Louis-Dupont** [comment]
Note: We agreed on adressing the changes concerning the denormalization in a following RP
